### PR TITLE
Make the CAReduce tests faster

### PR DESCRIPTION
**Pull request by abergeron**
This also fixes:
 - minor memory leak in the elemwise alloc code
 -  compile of ParamsType with CLinker of more than one op.

### CODE REVIEW

File: theano/gof/params_type.py [hunk 1]
```diff
@@ -684,6 +684,7 @@ def c_extract(self, name, sub, check_input=True):
         /* Seems c_init() is not called for a op param. So I call `new` here. */
         %(name)s = new %(struct_name)s;
 
+        { // This need a separate namespace for Clinker
```
Comment (nouiz) [inline theano/gof/params_type.py]: What is the consequence of this? a crash?

Comment (abergeron) [inline theano/gof/params_type.py]: A failure to compile.

**Comment by nouiz** [comment]
It would be great to have a test for that case.

**Comment by abergeron** [comment]
The optimizations include stuff that eliminates useless cases and merges some similar cases so we end up compiling less C code.

My timings indicate that test_c goes from ~200s to ~100s.

**Comment by abergeron** [comment]
What case?

**Comment by nouiz** [comment]
the failure to compile.

Can you confirm yout timing was always done with an empty cache?

**Comment by nouiz** [comment]
I approuve this PR. You can merge if CI passes (except for the timeout).

It would be great if you can add a test case for the crash fix in commit 3465a40

**Comment by abergeron** [comment]
Yes I made sure to delete the cache between each run.

**Comment by abergeron** [comment]
The new form of the test for CAReduce do test the CLinker case.  That's how I found the problem.

**Comment by nouiz** [comment]
jenkins fail.

**Comment by nouiz** [comment]
flake8: https://travis-ci.org/Theano/Theano/jobs/250419945#L907

**Comment by nouiz** [comment]
stoping travis.

**Comment by abergeron** [comment]
Someone killed almost all of the travis jobs for the last commit, why?

**Pull request merged**